### ISSUE TITLE
Support for streaming services

### ISSUE THREAD

**Issue by raamb** [singnet/snet-cli#245]
As we enable support for streaming services we need to decide how we will enable interaction with such services. 
Since the client generates stubs dynamically it needs to identify which service uses client side streaming, server side streaming or bi-directional streaming and accordingly capture inputs and render outputs.

cc @astroseger @arturgontijo @ferrouswheel  @vforvalerio87 @vsbogd

**Comment by ferrouswheel** [comment]
Happy to hear how others think it might work, but for me it's not clear how we can do this in a generic way for calling from the CLI without potential deadlocks, since the communication pattern isn't defined anywhere.

Streaming should probably be an advanced topic which service authors need to define the protocol for, and only supported when there is client code to handle the communication (i.e. the custom dapp UI components or the SDK).

**Comment by raamb** [comment]
> Happy to hear how others think it might work, but for me it's not clear how we can do this in a generic way for calling from the CLI without potential deadlocks, since the communication pattern isn't defined anywhere.

Yes, completely agree. This is a non-trivial problem and we need to have a way to deal with it i.e. we display an appropriate message where we cannot support and indicate options for the developer to try instead

**Comment by vsbogd** [comment]
We have discussed it with @astroseger today:
- if client uses SDK then it can implement protocol which is described by publisher in protobuf comments and there is no protocol specification problem;
- for snet-cli we can implement it asking client to provide number of messages to sent to the service and showing messages received one by one, after all snet-cli is low-level tool to make fast checks;
- it is a real problem for services which calls other services for this use case we need to find a way to describe API formally.

One of the ways to describe it formally would be a introducing:
```proto
message Chunk {
  bytes data = 1;
}
```
which can be used by service publisher to send big amount of data. If request message contains fields with Chunk type then SDK or snet-cli will break this bytes to the chunks and form stream of messages. Like:
```proto
message Request {
  string mime = 1;
  Chunk video = 2;
}
```
will be converted to the stream of Requests each contains only chunk of `video` field data.

**Comment by astroseger** [comment]
So in any case we need to have some predefined scenarios. Two the most simplest one
* **stream-as-request**, simple one time request-response scenario, but client sends request as a stream and wait response
* **stream-as-request**, simple one time request-response scenario, but client receives response as a stream. 
* other more complicated scenarios could be defined later if needed... 

We only need to way for the client to know which scenario he need to follow. For example as @vsbogd has proposed it can be simple "inheritance" of predefined (by us) types...

**Comment by ferrouswheel** [comment]
If we do use predefined types let's follow the way many grpc authors do it
and have a `one_of` stanza that splits every message into being a header or
Chunk, to avoid sending extra metadata e.g. mimetype with every streamed
message.
